File: BOJ/그래프/세그먼트 트리(Segment Tree)/2042_구간 합 구하기.cpp
#include <iostream>
#include <vector>
using namespace std;
int n, m, k;
vector<long long> tree;
vector<long long> arr;
void update(int idx, long long val);
long long sum(int idx);
int main(void) {
	scanf("%d%d%d", &n, &m, &k);
	tree.resize(n + 1);
	arr.resize(n + 1);
	for (int i = 1; i <= n; i++) {
		scanf("%lld", &arr[i]);
		update(i, arr[i]);
	}
	int cnt = m + k;
	while (cnt--) {
		int a;
		scanf("%d", &a);
		if (a == 1) {
			int b;
			long long c;
			scanf("%d%lld", &b, &c);
			long long diff = c - arr[b];
			arr[b] = c;
			update(b, diff);
		}
		else {
			int b, c;
			scanf("%d%d", &b, &c);
			printf("%lld\n", sum(c) - sum(b - 1));
		}
	}
}
void update(int idx, long long val) {
	while (idx <= n) {
		tree[idx] += val;
		idx += (idx & -idx);
	}
}
long long sum(int idx) {
	long long res = 0;
	while (idx > 0) {
		res += tree[idx];
		idx -= (idx & -idx);
	}
	return res;
}

File: BOJ/그래프/펜윅 트리(Fenwick Tree)/2042_구간 합 구하기.java
import java.io.*;
import java.util.StringTokenizer;

public class 2042_구간 합 구하기 {

    public static int n;
    public static int m;
    public static int k;
    public static long[] arr;
    public static long[] tree;

    public static void update(int idx, long value){
        while(idx<=n){
            tree[idx] += value;
            idx +=( idx&(-idx));
        }
    }

    public static long sum(int idx){
        long res = 0;
        while(idx > 0){
            res +=  tree[idx];
            idx -= (idx&(-idx));
        }
        return res;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());
        arr = new long[n+1];
        tree = new long[n+1];
        for(int i=1; i<=n;i++){
            st = new StringTokenizer(br.readLine());
            long value = Long.parseLong(st.nextToken());
            update(i,value);
            arr[i] = value;
        }
        for(int i=1; i<=m+k;i++){
            int a,b;
            st = new StringTokenizer(br.readLine());
            a = Integer.parseInt(st.nextToken());
            b = Integer.parseInt(st.nextToken());
            if(a==1){
                long c = Long.parseLong(st.nextToken());
                long diff = c - arr[b];
                update(b, diff);
                arr[b] = c;
            }
            else {
                int c = Integer.parseInt(st.nextToken());
                long res = sum(c) -sum(b-1);
                bw.write(String.valueOf(res)+"\n");
            }
        }
        bw.flush();
        br.close();
        bw.close();
    }
}